### PR TITLE
okhttp supports custom protocols. gh-809

### PR DESCRIPTION
**Pull request by galaxy-sea**
hello @OlgaMaciaszek , Complete the `OkHttpClient` custom protocol, but `Client.Default` and `ApacheHttp5Client` do not support http2.

### CODE REVIEW

File: spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/protocol/FeignOkProtocolsTests.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+class FeignOkProtocolsTests {
+
+	@Autowired
+	private Client feignClient;
+
+	@Autowired
+	private Application.ProtocolClient protocolClient;
+
+	@Test
+	void testFeignClientType() {
```
[the remark this answers is not in the document]
Comment (OlgaMaciaszek) [reply spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/protocol/FeignOkProtocolsTests.java:48]: Please add an assertion to verify that the client has the `Protocol.H2_PRIOR_KNOWLEDGE` present.

**Review by OlgaMaciaszek — CHANGES_REQUESTED** [review]
Hello, @galaxy-sea. Actually,  HTTP2 is supported by OkHttp and it's [enabled by default](https://github.com/square/okhttp/blob/5402ebaaa67ae05975e7d02490ef4003fdf3a4da/okhttp/src/jvmMain/kotlin/okhttp3/OkHttpClient.kt#L1144). However, this PR is still useful. I have added comments - please address them.

Also, there's [support in Apache HC5](https://hc.apache.org/httpcomponents-core-5.1.x/current/httpcore5-h2/apidocs/). We should add the support for it within gh-809 (can be in a separate PR) - will you work on that as well?

File: spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/FeignOkHttpConfigurationTests.java [hunk 2]
```diff
@@ -71,6 +75,13 @@ void shouldConfigureReadTimeout() {
 		assertThat(httpClient.readTimeoutMillis()).isEqualTo(9000);
 	}
 
+	@Test
+	void shouldProtocols() {
```
[the remark this answers is not in the document]
Comment (galaxy-sea) [reply spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/FeignOkHttpConfigurationTests.java]: done.

File: spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/FeignOkHttpConfigurationTests.java [hunk 3]
```diff
@@ -71,6 +75,13 @@ void shouldConfigureReadTimeout() {
...
+		OkHttpClient httpClient = context.getBean(OkHttpClient.class);
+
+		assertThat(httpClient.protocols()).isEqualTo(List.of(Protocol.H2_PRIOR_KNOWLEDGE));
```
[the remark this answers is not in the document]
Comment (galaxy-sea) [reply spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/FeignOkHttpConfigurationTests.java]: done.

File: spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/protocol/FeignOkProtocolsTests.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (galaxy-sea) [reply spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/protocol/FeignOkProtocolsTests.java]: done.

File: spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/protocol/Application.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.context.annotation.Import;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+/**
+ * @author changjin wei(魏昌进)
+ */
+public class Application {
```
[the remark this answers is not in the document]
Comment (galaxy-sea) [reply spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/protocol/Application.java]: done.

File: spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/protocol/FeignOkProtocolsTests.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+		assertThat(feignClient).isInstanceOf(FeignBlockingLoadBalancerClient.class);
+		FeignBlockingLoadBalancerClient client = (FeignBlockingLoadBalancerClient) feignClient;
+		Client delegate = client.getDelegate();
+		assertThat(delegate).isInstanceOf(feign.okhttp.OkHttpClient.class);
+	}
+
+	@Test
```
[the remark this answers is not in the document]
Comment (galaxy-sea) [reply spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/protocol/FeignOkProtocolsTests.java]: done. You're right.  I have deleted the relevant test code.

**Review by OlgaMaciaszek — CHANGES_REQUESTED** [review]
Thanks, @galaxy-sea. Have just added two cosmetic comments. Once this is done, we'll be able to merge it.

File: spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/protocol/FeignOkProtocolsTests.java [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+		ReflectionUtils.makeAccessible(field);
+		return ReflectionUtils.getField(field, target);
+	}
+
+	@Configuration(proxyBeanMethods = false)
+	@EnableAutoConfiguration
+	@RestController
+	@LoadBalancerClients
+	@Import(NoSecurityConfiguration.class)
+	public static class Application {
```
[the remark this answers is not in the document]
Comment (galaxy-sea) [reply spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/protocol/FeignOkProtocolsTests.java:76]: Expected `Expected protected static class`

File: spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/protocol/FeignOkProtocolsTests.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+
+	@Test
+	void shouldCreateCorrectFeignClientBeanWithProtocolFromProperties() {
+		assertThat(feignClient).isInstanceOf(FeignBlockingLoadBalancerClient.class);
+		FeignBlockingLoadBalancerClient client = (FeignBlockingLoadBalancerClient) feignClient;
+		Client delegate = client.getDelegate();
+		assertThat(delegate).isInstanceOf(feign.okhttp.OkHttpClient.class);
+		okhttp3.OkHttpClient OkHttpClient = (okhttp3.OkHttpClient) getField(delegate, "delegate");
+		assertThat(OkHttpClient.protocols()).containsExactly(Protocol.H2_PRIOR_KNOWLEDGE);
+
```
[the remark this answers is not in the document]
Comment (galaxy-sea) [reply spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/protocol/FeignOkProtocolsTests.java:62]: done.

**Review by OlgaMaciaszek — COMMENTED** [review]
Thanks, @galaxy-sea, I see the build is failing on checkstyle:
`/home/runner/work/spring-cloud-openfeign/spring-cloud-openfeign/spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/FeignOkHttpConfigurationTests.java:20:1: 'javax.net.ssl.HostnameVerifier' should be separated from previous imports. `

File: spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/FeignOkHttpConfigurationTests.java [hunk 8]
```diff
@@ -17,10 +17,10 @@
 package org.springframework.cloud.openfeign;
 
 import java.lang.reflect.Field;
-
```
Comment (OlgaMaciaszek) [inline spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/FeignOkHttpConfigurationTests.java:20]: This is what is causing the build checkstyle failure - please revert this line deletion.

**Review by OlgaMaciaszek — CHANGES_REQUESTED** [review]
Thanks, @galaxy-sea, the build is failing on checkstyle:
`/home/runner/work/spring-cloud-openfeign/spring-cloud-openfeign/spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/FeignOkHttpConfigurationTests.java:20:1: 'javax.net.ssl.HostnameVerifier' should be separated from previous imports. `. Have added a comment - please fix.

File: spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/FeignOkHttpConfigurationTests.java (hunk 8 above)
Comment (galaxy-sea) [reply spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/FeignOkHttpConfigurationTests.java]: Sorry, I resubmitted the code

Comment (OlgaMaciaszek) [reply spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/FeignOkHttpConfigurationTests.java]: No problem, thanks.

**Comment by OlgaMaciaszek** [comment]
@galaxy-sea I see there are some test failures that I would not expect as a result of your PR. Can you first merge changes from `origin/main` into your PR to get the updated artifact versions? If it still fails then, will take a closer look.

**Comment by galaxy-sea** [comment]
> @galaxy-sea I see there are some test failures that I would not expect as a result of your PR. Can you first merge changes from `origin/main` into your PR to get the updated artifact versions? If it still fails then, will take a closer look.

I submit

**Comment by OlgaMaciaszek** [comment]
Ok, @galaxy-sea - I'll take a look at it on Monday.

**Comment by OlgaMaciaszek** [comment]
Hi, @galaxy-sea, this is happening because the `Protocol` class has been compiled without `- parameters`, which is necessary for AOT processing of the properties file. I have [modified it to use Strings instead](84ebabce1ddafeaef1e40d5ee529b9ff54d84cf7). Merged within gh-825. Thanks for your work on this issue @galaxy-sea.

**Pull request closed without merging**